### PR TITLE
Asset: Fix assignment errors, remove usage of GET variables

### PR DESCRIPTION
**Pull request by trendspotter**
A PR to fix issue reported in #1405

Basically, the variables formerly passed via GET are now implied (hardcoded) based on the method used
- `setbase` = `/asset/asset/1/log/create?status=1`
- `assignperson` = `/asset/asset/1/log/create?status=2&type=person`
- `assignsite` = `/asset/asset/1/log/create?status=2&type=site`
- `assignorg` = `/asset/asset/1/log/create?status=2&type=organisation`

The `type` variable is removed altogether as the intended relation type is now inferred from the method. This also allowed a slight change in the condition logic in `asset_log_onaccept()` where the person assignment is now the default option in `ASSET_LOG_ASSIGN` branch unless explicitly stated otherwise via appropriate method.

I'm not aware of any other module where custom methods would be use in the same fashion, so any advice on how to make it prettier and more obvious are welcome.

### CODE REVIEW

File: modules/s3db/asset.py [hunk 1]
```diff
@@ -700,8 +700,17 @@ def asset_log_onaccept(form):
         """
 
         request = current.request
-        get_vars = request.get_vars
-        status = get_vars.get("status", None)
+
+        # Custom methods to avoid passing GET variables
+        method = current.response.s3.asset_log_method
+        print(method)
```
Comment (nursix) [inline modules/s3db/asset.py]: Debug code left in

**Comment by nursix** [comment]
I think this works well eliminating the vulnerability caused by GET vars inspection far away from the controller, and prettifying further goes beyond the problem at hand.

Passing parameters from the controller to the onaccept via response.s3 is fragile (still better than inspecting GET vars in the onaccept, though), and requires proper documentation, which you have done.

What's really wrong here is that the entity type of the assignment can't be inferred from the asset log record, which is a glaring weakness of the data model. So, any further improvement should try to solve that, so that the onaccept can inspect the record instead of requiring a hint from the controller.

Aliasing/overriding URL methods, on the other hand, is absolutely a valid pattern - we're using that elsewhere (e.g. deploy), it is allowed and intended in the framework.

**Comment by nursix** [comment]
...and when I say "glaring weakness" then I mean it: 

There is no way to reconstruct the current or any historic status of the asset from the log - because you always need the "type" (explicit or implicit) from the request that created the log entry, which is lost as soon as the call is over.

That's quite a concept flaw if you ask me - what kind of a "log" is that if you can't reconstruct the current/historic statuses from the trail of log entries?

So if you really wanted to fix it, then you should add the entity type to the log record - and then make the onaccept inspect the log trail to determine status and location of the asset.

**Comment by nursix** [comment]
Imagine the situation where the assignment happens in an offline instance, and then you sync two or three log entries to the master - how can you determine the current assignment type?

Or, how can the log answer the question to what or whom the asset was assigned at time X somewhere in the past (and I mean: that's the whole point with a log, isn't it?).

One option could be to assign via super-entities (=key tables), i.e. pe_id and site_id. Then, we apply the simple yet intuitive logic that when no site_id is set, the asset is to be tracked with with the person entity identified by the pe_id (organisation, person, team via pe_id) - otherwise it is located at the site specified by site_id.

**Comment by trendspotter** [comment]
Debug code removed, comments slightly adjusted to be a bit more clear (there's no point of mentioning GET variables if the reader doesn't know the history behind it)

> So if you really wanted to fix it, then you should add the entity type to the log record - and then make the onaccept inspect the log trail to determine status and location of the asset.

That's kinda is what I was thinking by [option 2](https://github.com/sahana/eden/issues/1405#issuecomment-335022906) - ie. not to explicitly *require* any of the values by data model based on which method was used, but instead use generic form and check which fields did the user filled in. If they filled in person, then assign to person, if they filled in site, then assign to site etc. I imagine this may get more complicated when the user sets eg. an organization AND person who has nothing to do with the org, but it *should* still be fairly easy to catch in data model validation.

Anyway I think there's more history behind the assignments then it's visible from the code. E.g. the disabled `ASSET_LOG_RETURN`, so ultimately, some deeper modifications may be in order. But that's stage 2. Stage 1 is to make the things work without errors so we can see what they were supposed to do. And once they do, we can ~~break them again~~ improve them.

**Comment by nursix** [comment]
Sure Sure - you were asking for suggestions how this could be made "prettier", so that is why I made suggestions.

Apart from that, I think this PR resolves the original issue - so stage 1 is good to go ;)

File: modules/s3db/asset.py [hunk 2]
```diff
@@ -1030,17 +1043,16 @@ def asset_log_prep(r):
         table.site_id.requires = IS_ONE_OF(db, "org_site.site_id",
                                            table.site_id.represent)
 
-    elif status == ASSET_LOG_RETURN:
-        crud_strings.msg_record_created = T("Returned")
-        table.person_id.label = T("Returned From")
-        table.person_id.default = current_log.person_id
-        table.site_id.readable = False
-        table.site_id.writable = False
+    # Return functionality removed  - as it doesn't set site_id & organisation_id in the logs
```
Comment (flavour) [inline modules/s3db/asset.py]: I'm not sure this is a good enough reason to remove it completely?
Surely site_id & organisation_id don't need logging anyway, as these are now back to the original for the asset?

File: modules/s3db/asset.py [hunk 3]
```diff
@@ -700,8 +700,16 @@ def asset_log_onaccept(form):
         """
 
         request = current.request
-        get_vars = request.get_vars
-        status = get_vars.get("status", None)
+
+        # Custom methods to allow form customization for specific cases
+        method = current.response.s3.asset_log_method
```
Comment (flavour) [inline modules/s3db/asset.py]: I would add a little note:
Read method set in prep

File: modules/s3db/asset.py [hunk 4]
```diff
@@ -982,6 +985,20 @@ def asset_log_prep(r):
 
     table = db.asset_log
 
+    # Custom methods to allow form customization for specific cases
+    # Also passed via response.s3.asset_log_method to asset_log_onaccept()
```
Comment (flavour) [inline modules/s3db/asset.py]: Lovely :)

**Comment by flavour** [comment]
Thanks for this - some definite enhancements here :)
Please don't remove the ASSET_LOG_RETURN though & a suggestion for slight enhancement of docs.

**Comment by trendspotter** [comment]
That `ASSET_LOG_RETURN` comment has been just copied from [here](https://github.com/sahana/eden/blob/c1da768eb804d55cdc988512cadeeeb650b0ab98/modules/s3db/asset.py#L1143) as the functionality is anyway disabled there. I've disabled it even in the `elif` to not confuse the next person reading that code. The `status` is now set either via custom method or via the form and there are no circumstances under which it could be  `ASSET_LOG_RETURN` . Anyway, I've uncommented it again, but It'll be just another piece of dead code.

Added the comment about `prep` in `onaccept`.

**Pull request merged**